cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -792,60 +818,81 @@ namespace cooperative_lanechange
         }
     }
 
-    int CooperativeLaneChangePlugin::getNearestRouteIndex(lanelet::BasicLineString2d& points, const cav_msgs::VehicleState& state) const
-    {
-        lanelet::BasicPoint2d veh_point(state.X_pos_global, state.Y_pos_global);
-                double min_distance = std::numeric_limits<double>::max();
-        int i = 0;
+    int CooperativeLaneChangePlugin::get_nearest_route_point_index(lanelet::BasicLineString2d& points, double ending_downtrack){
         int best_index = 0;
-        for (const auto& p : points)
-        {
-            double distance = lanelet::geometry::distance2d(p,veh_point);
-            if (distance < min_distance)
-            {
-            best_index = i;
-            min_distance = distance;
+        bool set_index = false;
+        for(int i=0;i < points.size();i++){
+            double downtrack = wm_->routeTrackPos(points[i]).downtrack;
+            if(downtrack > ending_downtrack){
+                best_index = i - 1;
+                ROS_DEBUG_STREAM("get_nearest_route_point_index>> Found best_idx: " << best_index<<", points[i].x(): " << points[best_index].x() << ", points[i].y(): " << points[best_index].y() << ", downtrack: "<< downtrack);
+                set_index = true;
+                break;
             }
-            i++;
         }
-        return best_index;
+        if (!set_index)
+        {
+            best_index = points.size() - 1;
+        }
+        ROS_DEBUG_STREAM("get_nearest_route_point_index>> Found best_idx: " << best_index<<", points[i].x(): " << points[best_index].x() << ", points[i].y(): " << points[best_index].y());
 
```

Comment:
If you initialize best_index as points.size() -1, you can avoid having this if(!set_index) loop

Reply:
Just a thought, would it make sense to call this function, get_nearest_point_index as well?

Reply:
That's a good idea. fixed best_index and renamed the function